### PR TITLE
Add rewrite for semantic constants.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
First, I don't expect this to be merged!

What this is is a pass over the IR that occurs before dead branch pruning (and before typing!) which rewrites values which are known constants through their semantics as `ir.Const` nodes. An example might be replacing a `getattr(ndarray_instance, 'ndim')` on an input argument with the constant value of `ndim`.

Some examples:
```python
@jit(nopython=True)
def test_1(array):
    if array.ndim == 2:
        if array.shape[1] == 2:
            return 1
    return 0

array = np.empty((1,))
test_1(array)
```
The above would normally fail as the true branch would need to be typed irrespective of whether it is executed, and as `array.shape[1]` would be rewritten as a `static_getitem` and the input array does not have a `shape[1]` index (OOB) then compilation fails (this is issue https://github.com/numba/numba/issues/4108). With the PR code added, the `array.ndim` gets rewritten as an `ir.Const(1)` and as a result `dead_branch_pruning` kills the true branch, the branch body is removed from the IR and typing/general execution can proceed.

Other examples are:
```python
@jit(nopython=True)
def test_2(tup):
    if len(tup) == 3:
        if tup[2] == 2:
            return 1
    return 0

test_2((1,))
```
In `test_2` the `len(tup)` call gets rewritten to the length of `tup`, `ir.Const(1)` and true branch is subsequently pruned.

```python
@jit(nopython=True)
def test_3(arr):
    if arr.dtype == np.float64:
        x = 1
    else:
        x = 10
    return x

test_3(np.zeros((10,)))
```
In the above `arr.dtype` is rewritten as a `ir.Const(types.float64)` and `np.float64` is also rewritten as the same, hence the false branch can be pruned.

For completeness, here's the before/after IR for each of the above functions:
```
-----------------------rewrite_semantic_constants: test_1-----------------------
*************************************before*************************************
label 0:
    array = arg(0, name=array)               ['array']
    $0.2 = getattr(value=array, attr=ndim)   ['$0.2', 'array']
    $const0.3 = const(int, 2)                ['$const0.3']
    $0.4 = $0.2 == $const0.3                 ['$0.2', '$0.4', '$const0.3']
    del $const0.3                            []
    del $0.2                                 []
    branch $0.4, 10, 28                      ['$0.4']
label 10:
    del $0.4                                 []
    $10.2 = getattr(value=array, attr=shape) ['$10.2', 'array']
    del array                                []
    $const10.3 = const(int, 1)               ['$const10.3']
    $10.4 = static_getitem(value=$10.2, index=1, index_var=$const10.3) ['$10.2', '$10.4', '$const10.3']
    del $const10.3                           []
    del $10.2                                []
    $const10.5 = const(int, 2)               ['$const10.5']
    $10.6 = $10.4 == $const10.5              ['$10.4', '$10.6', '$const10.5']
    del $const10.5                           []
    del $10.4                                []
    branch $10.6, 24, 28                     ['$10.6']
label 24:
    del $10.6                                []
    $const24.1 = const(int, 1)               ['$const24.1']
    $24.2 = cast(value=$const24.1)           ['$24.2', '$const24.1']
    del $const24.1                           []
    return $24.2                             ['$24.2']
label 28:
    del array                                []
    del $10.6                                []
    del $0.4                                 []
    $const28.1 = const(int, 0)               ['$const28.1']
    $28.2 = cast(value=$const28.1)           ['$28.2', '$const28.1']
    del $const28.1                           []
    return $28.2                             ['$28.2']
*************************************after**************************************
label 0:
    array = arg(0, name=array)               ['array']
    $0.2 = const(int, 1)                     ['$0.2']
    $const0.3 = const(int, 2)                ['$const0.3']
    $0.4 = $0.2 == $const0.3                 ['$0.2', '$0.4', '$const0.3']
    del $const0.3                            []
    del $0.2                                 []
    branch $0.4, 10, 28                      ['$0.4']
label 10:
    del $0.4                                 []
    $10.2 = getattr(value=array, attr=shape) ['$10.2', 'array']
    del array                                []
    $const10.3 = const(int, 1)               ['$const10.3']
    $10.4 = static_getitem(value=$10.2, index=1, index_var=$const10.3) ['$10.2', '$10.4', '$const10.3']
    del $const10.3                           []
    del $10.2                                []
    $const10.5 = const(int, 2)               ['$const10.5']
    $10.6 = $10.4 == $const10.5              ['$10.4', '$10.6', '$const10.5']
    del $const10.5                           []
    del $10.4                                []
    branch $10.6, 24, 28                     ['$10.6']
label 24:
    del $10.6                                []
    $const24.1 = const(int, 1)               ['$const24.1']
    $24.2 = cast(value=$const24.1)           ['$24.2', '$const24.1']
    del $const24.1                           []
    return $24.2                             ['$24.2']
label 28:
    del array                                []
    del $10.6                                []
    del $0.4                                 []
    $const28.1 = const(int, 0)               ['$const28.1']
    $28.2 = cast(value=$const28.1)           ['$28.2', '$const28.1']
    del $const28.1                           []
    return $28.2                             ['$28.2']
--------------------------------------------------------------------------------
-----------------------rewrite_semantic_constants: test_2-----------------------
*************************************before*************************************
label 0:
    tup = arg(0, name=tup)                   ['tup']
    $0.1 = global(len: <built-in function len>) ['$0.1']
    $0.3 = call $0.1(tup, func=$0.1, args=[Var(tup, issue4108.py (16))], kws=(), vararg=None) ['$0.1', '$0.3', 'tup']
    del $0.1                                 []
    $const0.4 = const(int, 3)                ['$const0.4']
    $0.5 = $0.3 == $const0.4                 ['$0.3', '$0.5', '$const0.4']
    del $const0.4                            []
    del $0.3                                 []
    branch $0.5, 12, 28                      ['$0.5']
label 12:
    del $0.5                                 []
    $const12.2 = const(int, 2)               ['$const12.2']
    $12.3 = static_getitem(value=tup, index=2, index_var=$const12.2) ['$12.3', '$const12.2', 'tup']
    del tup                                  []
    del $const12.2                           []
    $const12.4 = const(int, 2)               ['$const12.4']
    $12.5 = $12.3 == $const12.4              ['$12.3', '$12.5', '$const12.4']
    del $const12.4                           []
    del $12.3                                []
    branch $12.5, 24, 28                     ['$12.5']
label 24:
    del $12.5                                []
    $const24.1 = const(int, 1)               ['$const24.1']
    $24.2 = cast(value=$const24.1)           ['$24.2', '$const24.1']
    del $const24.1                           []
    return $24.2                             ['$24.2']
label 28:
    del tup                                  []
    del $12.5                                []
    del $0.5                                 []
    $const28.1 = const(int, 0)               ['$const28.1']
    $28.2 = cast(value=$const28.1)           ['$28.2', '$const28.1']
    del $const28.1                           []
    return $28.2                             ['$28.2']
*************************************after**************************************
label 0:
    tup = arg(0, name=tup)                   ['tup']
    $0.1 = global(len: <built-in function len>) ['$0.1']
    $0.3 = const(int, 1)                     ['$0.3']
    del $0.1                                 []
    $const0.4 = const(int, 3)                ['$const0.4']
    $0.5 = $0.3 == $const0.4                 ['$0.3', '$0.5', '$const0.4']
    del $const0.4                            []
    del $0.3                                 []
    branch $0.5, 12, 28                      ['$0.5']
label 12:
    del $0.5                                 []
    $const12.2 = const(int, 2)               ['$const12.2']
    $12.3 = static_getitem(value=tup, index=2, index_var=$const12.2) ['$12.3', '$const12.2', 'tup']
    del tup                                  []
    del $const12.2                           []
    $const12.4 = const(int, 2)               ['$const12.4']
    $12.5 = $12.3 == $const12.4              ['$12.3', '$12.5', '$const12.4']
    del $const12.4                           []
    del $12.3                                []
    branch $12.5, 24, 28                     ['$12.5']
label 24:
    del $12.5                                []
    $const24.1 = const(int, 1)               ['$const24.1']
    $24.2 = cast(value=$const24.1)           ['$24.2', '$const24.1']
    del $const24.1                           []
    return $24.2                             ['$24.2']
label 28:
    del tup                                  []
    del $12.5                                []
    del $0.5                                 []
    $const28.1 = const(int, 0)               ['$const28.1']
    $28.2 = cast(value=$const28.1)           ['$28.2', '$const28.1']
    del $const28.1                           []
    return $28.2                             ['$28.2']
--------------------------------------------------------------------------------
-----------------------rewrite_semantic_constants: test_3-----------------------
*************************************before*************************************
label 0:
    arr = arg(0, name=arr)                   ['arr']
    $0.2 = getattr(value=arr, attr=dtype)    ['$0.2', 'arr']
    del arr                                  []
    $0.3 = global(np: <module 'numpy' from '<env>/lib/python3.7/site-packages/numpy/__init__.py'>) ['$0.3']
    $0.4 = getattr(value=$0.3, attr=float64) ['$0.3', '$0.4']
    del $0.3                                 []
    $0.5 = $0.2 == $0.4                      ['$0.2', '$0.4', '$0.5']
    del $0.4                                 []
    del $0.2                                 []
    branch $0.5, 12, 18                      ['$0.5']
label 12:
    del $0.5                                 []
    $const12.1 = const(int, 1)               ['$const12.1']
    x = $const12.1                           ['$const12.1', 'x']
    del $const12.1                           []
    jump 22                                  []
label 18:
    del $0.5                                 []
    $const18.1 = const(int, 10)              ['$const18.1']
    x = $const18.1                           ['$const18.1', 'x']
    del $const18.1                           []
    jump 22                                  []
label 22:
    $22.2 = cast(value=x)                    ['$22.2', 'x']
    del x                                    []
    return $22.2                             ['$22.2']
*************************************after**************************************
label 0:
    arr = arg(0, name=arr)                   ['arr']
    $0.2 = const(Float, float64)             ['$0.2']
    del arr                                  []
    $0.3 = global(np: <module 'numpy' from '<env>/lib/python3.7/site-packages/numpy/__init__.py'>) ['$0.3']
    $0.4 = const(Float, float64)             ['$0.4']
    del $0.3                                 []
    $0.5 = $0.2 == $0.4                      ['$0.2', '$0.4', '$0.5']
    del $0.4                                 []
    del $0.2                                 []
    branch $0.5, 12, 18                      ['$0.5']
label 12:
    del $0.5                                 []
    $const12.1 = const(int, 1)               ['$const12.1']
    x = $const12.1                           ['$const12.1', 'x']
    del $const12.1                           []
    jump 22                                  []
label 18:
    del $0.5                                 []
    $const18.1 = const(int, 10)              ['$const18.1']
    x = $const18.1                           ['$const18.1', 'x']
    del $const18.1                           []
    jump 22                                  []
label 22:
    $22.2 = cast(value=x)                    ['$22.2', 'x']
    del x                                    []
    return $22.2                             ['$22.2']
--------------------------------------------------------------------------------
```

**Comment by luk-f-a** [comment]
@stuartarchibald, this looks useful, out of curiosity, any reason why you wouldn't merge it?

**Comment by seibert** [comment]
I think he means that this is an experiment to get feedback on the idea. This PR as-is may not be merge, but a future PR based on this one may be created after some review.

**Comment by luk-f-a** [comment]
thanks @seibert, I get it now. I am getting fascinated by how many things can be done by JIT compilation, in terms of code specialization. It seems to me that it is possible to have highly dynamic code which half-way through, once all the variables are known, can compile a very efficient of itself and keep running. I guess this is something that Pypy attempts to do as a tracing JIT, but Numba allows users to choose when to switch from dynamic Python code and efficient compiled code. Specifically about this PR, @stuartarchibald , what is the situation of closure variables? Are they being constant-propagated at the moment? If not, would this PR help? I'm thinking about, for example, dict_closure[constant_key]? That could be replaced by a constant.

**Comment by sklam** [comment]
This is a useful feature to have and it'd be be even better if our type-inference can feed typeinfo to it.  Currently, this rewrite pass has to happen before type-inference.

**Comment by stuartarchibald** [comment]
Following core dev discussion, we decided this is a good thing to have right now as it fixes the problem, it also doesn't preclude future efforts on partial/incomplete typing from helping to do a better job. A variant of this pass that is typed would also seed the IR with a load of constants, potentially allowing a second branch prune to take place and help with e.g. future work on constant propagation.

TODO:
- [ ] tests
- [ ] comments and docstrings
- [ ] check what, if anything, this might break in ParallelAccelerator

File: numba/analysis.py [hunk 1]
```diff
@@ -434,3 +434,65 @@ def resolve_input_arg_const(input_arg_idx):
...
+        repl_idx = defns.index(val)
+        defns[repl_idx] = stmt.value
+
+    from .ir_utils import get_definition, guard
+    for blk in func_ir.blocks.values():
+        for stmt in blk.body:
+            if isinstance(stmt, ir.Assign):
+                val = stmt.value
+                if isinstance(val, ir.Expr):
+                    if getattr(val, 'op', None) == 'getattr':
```
Comment (sklam) [inline numba/analysis.py]: The function has too many levels.  I'd suggest lifting this branch into a `rewrite_array_ndim()` and the next `if` into a `rewrite_tuple_len()`.

File: numba/tests/test_analysis.py [hunk 2]
```diff
@@ -551,3 +558,55 @@ def impl(array, x=None, a=None):
...
+                    return 1
+            else:
+                return 10
+
+        self.assert_prune(impl, (types.Array(types.float64, 2, 'C'),), [False,
+                                                                        None],
+                          np.zeros((2, 3)))
+        self.assert_prune(impl, (types.Array(types.float64, 1, 'C'),), [True,
+                                                                        'both'],
+                          np.zeros((2,)))
```
Comment (sklam) [inline numba/tests/test_analysis.py]: Can you add another case that checks that pruning doesn't happen when a non-array type is used?

i.e.
```python
from numba import njit
import collections
import numpy as np

FakeArray = collections.namedtuple('FakeArray', ['ndim'])

fa = FakeArray(ndim=2)

@njit
def foo(fa):
    if fa.ndim == 2:
        return fa.ndim
    else:
        object()
       
    
print(foo(np.ones((2,2))))
foo(fa)  # <--- this fails
```

File: numba/analysis.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/analysis.py]: Good point, done.

File: numba/tests/test_analysis.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/tests/test_analysis.py]: I think I've addressed this now in 2aa5e3e, please let me know what you think, thanks!

**Pull request merged**